### PR TITLE
migrate to mackerel.Client

### PR DESCRIPTION
**Pull request by lufia**
I migrated **mackerel-agent/mackerel** to use *mackerel.Client*.

### CODE REVIEW

File: mackerel/check.go [hunk 1]
```diff
@@ -33,9 +33,7 @@ func (api *API) ReportCheckMonitors(hostID string, reports []*checks.Report) err
 			MaxCheckAttempts:     normalize(report.MaxCheckAttempts, 0),
 		}
 	}
-	resp, err := api.postJSON("/api/v0/monitoring/checks/report", payload)
-	defer closeResp(resp)
-	return err
+	return api.c.PostCheckReports(payload)
 }
```
Comment (lufia) [inline mackerel/check.go]: https://github.com/mackerelio/mackerel-client-go/blob/master/checks.go#L65

File: mackerel/api.go [hunk 2]
```diff
@@ -199,26 +199,8 @@ func (api *API) FindHostByCustomIdentifier(customIdentifier string) (*mkr.Host,
...
-	}
-	err = json.NewDecoder(resp.Body).Decode(&data)
-	if err != nil {
-		return "", err
-	}
-
-	return data.ID, nil
+func (api *API) CreateHost(hostParam *mkr.CreateHostParam) (string, error) {
+	return api.c.CreateHost(hostParam)
 }
```
Comment (lufia) [inline mackerel/api.go]: https://github.com/mackerelio/mackerel-client-go/blob/c64facc02ae71213da9b8a423b3cfeab3c3ea387/hosts.go#L216

File: mackerel/api.go [hunk 3]
```diff
@@ -204,14 +204,9 @@ func (api *API) CreateHost(hostParam *mkr.CreateHostParam) (string, error) {
...
 
 // UpdateHost updates the host information on Mackerel.
-func (api *API) UpdateHost(hostID string, hostParam mkr.UpdateHostParam) error {
-	resp, err := api.putJSON("/api/v0/hosts/"+hostID, hostParam)
-	defer closeResp(resp)
-	if err != nil {
-		return err
-	}
-
-	return nil
```
Comment (lufia) [inline mackerel/api.go]: https://github.com/mackerelio/mackerel-client-go/blob/c64facc02ae71213da9b8a423b3cfeab3c3ea387/hosts.go#L234

File: mackerel/metadata.go [hunk 4]
```diff
@@ -1,13 +1,6 @@
...
-import (
-	"net/http"
-)
-
 // PutMetadata creates or updates metadata of a host.
-func (api *API) PutMetadata(hostID string, namespace string, metadata interface{}) (*http.Response, error) {
-	resp, err := api.putJSON("/api/v0/hosts/"+hostID+"/metadata/"+namespace, metadata)
-	defer closeResp(resp)
-
-	return resp, err
```
Comment (lufia) [inline mackerel/metadata.go]: https://github.com/mackerelio/mackerel-client-go/blob/c64facc02ae71213da9b8a423b3cfeab3c3ea387/host_metadata.go#L72

File: command/command.go [hunk 5]
```diff
@@ -632,10 +630,10 @@ func NewMackerelClient(apibase, apikey, ver, rev string, verbose bool) (*mackere
 	if err != nil {
 		return nil, err
 	}
-	api.UA = buildUA(ver, rev)
-	api.DefaultHeaders = http.Header{}
-	api.DefaultHeaders.Add("X-Agent-Version", ver)
-	api.DefaultHeaders.Add("X-Revision", rev)
+	api.UserAgent = buildUA(ver, rev)
+	// We must not set api.DefaultHeaders; it is pointing to same value of api.c.AdditionalHeaders.
```
Comment (itchyny) [inline command/command.go]: api.c. does not exist, but api.Client does.

File: mackerel/api.go [hunk 6]
```diff
@@ -68,243 +48,30 @@ func infoError(msg string) *InfoError {
...
-		return api.UA
-	}
-	return "mackerel-agent/0.0.0"
+	c.AdditionalHeaders = make(http.Header)
+	// TODO(lufia): should we set a timeout explicitly?
+	//c.HTTPClient.Timeout = apiRequestTimeout
+	return &API{Client: c}, nil
 }
 
 var apiRequestTimeout = 30 * time.Second
```
Comment (itchyny) [inline mackerel/api.go]: Maybe this line can be deleted.

File: mackerel/api.go [hunk 7]
```diff
@@ -68,243 +48,30 @@ func infoError(msg string) *InfoError {
...
-}
-
-func (api *API) getUA() string {
-	if api.UA != "" {
-		return api.UA
-	}
-	return "mackerel-agent/0.0.0"
+	c.AdditionalHeaders = make(http.Header)
+	// TODO(lufia): should we set a timeout explicitly?
+	//c.HTTPClient.Timeout = apiRequestTimeout
```
Comment (itchyny) [inline mackerel/api.go]: Letting the default value of the client is ok imo.

File: command/command.go [hunk 8]
```diff
@@ -613,8 +612,7 @@ func (app *App) UpdateHostSpecs() {
 		return
 	}
 
-	err = app.API.UpdateHost(app.Host.ID, mkr.UpdateHostParam(hostParam))
-
+	_, err = app.API.UpdateHost(app.Host.ID, (*mkr.UpdateHostParam)(hostParam))
```
Comment (itchyny) [inline command/command.go]: This cast is required?

File: mackerel/api.go [hunk 9]
```diff
@@ -1,54 +1,34 @@
...
-}
-
-// Error represents API error
-type Error struct {
-	StatusCode int
-	Message    string
-}
-
-func (aperr *Error) Error() string {
-	return fmt.Sprintf("API error. status: %d, msg: %s", aperr.StatusCode, aperr.Message)
```
Comment (itchyny) [inline mackerel/api.go]: This is different from ApiError of mackerel-client-go so log message can change but maybe ok...

File: mackerel/api.go [hunk 10]
```diff
@@ -68,243 +48,30 @@ func infoError(msg string) *InfoError {
...
-			}
-		}
-	}
-	req.Header.Add("X-Api-Key", api.APIKey)
-	req.Header.Set("User-Agent", api.getUA())
-
-	if api.Verbose {
-		dump, err := httputil.DumpRequest(req, true)
-		if err == nil {
-			logger.Tracef("%s", dump)
```
Comment (itchyny) [inline mackerel/api.go]: Logger prefix `<api>` for log in verbose mode will be lost but maybe ok.

File: command/command.go (hunk 8 above)
Comment (lufia) [reply command/command.go]: Yes, this is required. hostParam is typed as `*mkr.CreateHostParam`.

**Comment by lufia** [comment]
I fixed lines that is commented.

File: mackerel/check.go [hunk 11]
```diff
@@ -33,9 +33,7 @@ func (api *API) ReportCheckMonitors(hostID string, reports []*checks.Report) err
 			MaxCheckAttempts:     normalize(report.MaxCheckAttempts, 0),
 		}
 	}
-	resp, err := api.postJSON("/api/v0/monitoring/checks/report", payload)
-	defer closeResp(resp)
-	return err
+	return api.Client.PostCheckReports(payload)
```
Comment (astj) [inline mackerel/check.go]: [nit]
`payload` might be named something like `reports` ?

File: go.sum [hunk 12]
```diff
@@ -34,6 +34,8 @@ github.com/mackerelio/golib v0.0.0-20190411032134-c87047ca454e h1:MwWzFCq2tENDy2
 github.com/mackerelio/golib v0.0.0-20190411032134-c87047ca454e/go.mod h1:kbqYA8VcFqcMt07v+GPSZQwqtwc7Wr0V0vzxUNMrk7E=
 github.com/mackerelio/mackerel-client-go v0.3.0 h1:4SFqVdKIWz5IiL/LY/SosoZJ73zDpgaX2rw0/JW6mLo=
```
Comment (astj) [inline go.sum]: Can you clean those old dependencies by something like `go mod tidy`?

File: mackerel/api.go (hunk 10 above)
Comment (lufia) [reply mackerel/api.go]: I've fixed at edde7d6

File: go.sum (hunk 12 above)
Comment (lufia) [reply go.sum]: I'm not sure. Perhaps I could remove *go.sum* and then regenerate it.

**Pull request merged**